### ISSUE TITLE
Type issue when extending SnackbarProvider Components with custom component

### ISSUE THREAD

**Issue by elisechant** [iamhosseindhv/notistack#592]
SnackbarProviderProps Components prop type does not support non standard component variant types.

## Expected Behavior

TypeScript compiler passes for extending SnackbarProvider as described in docs: 

```
<SnackbarProvider
    Components={{
      default: forwardRef(SnackbarItem),
      error: forwardRef(SnackbarItem),
      survey: forwardRef(SnackbarSurveyItem),    // <- custom variant
    }}
  >
    {children}
  </SnackbarProvider>
```

## Current Behavior

I'm using `SnackbarProvider` to provide custom UI for item variants "error", "warning", "info", "success" without issue. 

When I add the custom variant "survey", the code runs as expected, however I see a type error (strict mode):

````
ERROR in ./src/components/Snackbar/SnackbarProvider.tsx:16:7
TS2769: No overload matches this call.
  Overload 1 of 2, '(props: SnackbarProviderProps | Readonly<SnackbarProviderProps>): SnackbarProvider', gave the following error.
    Type '{ default: React.ForwardRefExoticComponent<Omit<SnackbarItemProps, "ref"> & React.RefAttributes<unknown>>; error: React.ForwardRefExoticComponent<Omit<SnackbarItemProps, "ref"> & React.RefAttributes<...>>; info: React.ForwardRefExoticComponent<...>; warning: React.ForwardRefExoticComponent<...>; success: React.Forwa...' is not assignable to type '{ info?: JSXElementConstructor<any> | undefined; success?: JSXElementConstructor<any> | undefined; error?: JSXElementConstructor<any> | undefined; default?: JSXElementConstructor<...> | undefined; warning?: JSXElementConstructor<...> | undefined; }'.
      Object literal may only specify known properties, and 'survey' does not exist in type '{ info?: JSXElementConstructor<any> | undefined; success?: JSXElementConstructor<any> | undefined; error?: JSXElementConstructor<any> | undefined; default?: JSXElementConstructor<...> | undefined; warning?: JSXElementConstructor<...> | undefined; }'.
  Overload 2 of 2, '(props: SnackbarProviderProps, context: any): SnackbarProvider', gave the following error.
    Type '{ default: React.ForwardRefExoticComponent<Omit<SnackbarItemProps, "ref"> & React.RefAttributes<unknown>>; error: React.ForwardRefExoticComponent<Omit<SnackbarItemProps, "ref"> & React.RefAttributes<...>>; info: React.ForwardRefExoticComponent<...>; warning: React.ForwardRefExoticComponent<...>; success: React.Forwa...' is not assignable to type '{ info?: JSXElementConstructor<any> | undefined; success?: JSXElementConstructor<any> | undefined; error?: JSXElementConstructor<any> | undefined; default?: JSXElementConstructor<...> | undefined; warning?: JSXElementConstructor<...> | undefined; }'.
      Object literal may only specify known properties, and 'survey' does not exist in type '{ info?: JSXElementConstructor<any> | undefined; success?: JSXElementConstructor<any> | undefined; error?: JSXElementConstructor<any> | undefined; default?: JSXElementConstructor<...> | undefined; warning?: JSXElementConstructor<...> | undefined; }'.
    14 |       warning: forwardRef(SnackbarItem),
    15 |       success: forwardRef(SnackbarItem),
  > 16 |       survey: forwardRef(SnackbarSurveyItem),
       |       ^^^^^^
````

Similarly, when I select the variant in a view:

````
TS2322: Type '"survey"' is not assignable to type '"info" | "success" | "error" | "default" | "warning" | undefined'.
    37 |   useEffect(() => {
    38 |     enqueueSnackbar('', {
  > 39 |       variant: 'survey',
       |       ^^^^^^^
    40 |       children: (props) => <SurveyNPS {...props} />,
    41 |       persist: true,
    42 |       preventDuplicate: true,
````

On inspection of library types, could it be that `Components` type is too narrow an only supports known Variants? 

````
export interface SnackbarProviderProps extends SharedProps {
  // ...
  Components?: {
    [variant in VariantType]?: React.JSXElementConstructor<any>;
  };
```

**Comment by marc-at-brightnight** [comment]
I have the same issue. Interestingly, if I hover over `VariantType` in my IDE, it seems to recognize the extra variants I've defined in my `notistack.d.ts` (`notifications` and `otherNotifications`).

![image](https://github.com/iamhosseindhv/notistack/assets/76454594/08b5e56d-06d9-471e-9798-4dd95654bd6b)

At this point, not really sure what else it would be.

Certainly would like to upgrade our repo to v3 but this is definitely a blocker. 

```
// notistack.d.ts

import { UINotification } from './types';

declare module 'notistack' {
  import { SnackbarKey, SnackbarMessage } from 'notistack';

  interface NotificationSnackProps {
    id: SnackbarKey;
    message: SnackbarMessage;
    notification: UINotification;
    handleClick: (notif: UINotification, redirect?: boolean) => void;
  }

  interface OtherNotificationsSnackProps {
    id: SnackbarKey;
    message: SnackbarMessage;
    handleClick: () => void;
    handleClearAll: () => void;
  }

  interface VariantOverrides {
    // "extra" props it takes in options of `enqueueSnackbar`
    notification: NotificationSnackProps;
    otherNotifications: OtherNotificationsSnackProps;
  }
}
```

**Comment by iamhosseindhv** [comment]
@elisechant When you define a custom variant, you need define it as explained in the docs: https://notistack.com/features/customization#custom-variant-(typescript)

TLDR:
```ts
declare module 'notistack' {
  interface VariantOverrides {
     // ... survey variant with its custom props
      survey: {
        
      }
      // or simply:
      survey: true
  }
}

```

Let me know if you're still facing issues

**Comment by elisechant** [comment]
Thank you @iamhosseindhv, apologies for missing that.

**Comment by marc-at-brightnight** [comment]
Excuse me @iamhosseindhv sir, but what am I missing? I have a `notistack.d.ts`. I added the `src/*.d.ts` to my `tsconfig.json` but still have the same error. Using Typescript 5.1.6 and notistack 3.0.1

```
src/App.tsx:252:15 - error TS2769: No overload matches this call.
  Overload 1 of 2, '(props: SnackbarProviderProps | Readonly<SnackbarProviderProps>): SnackbarProvider', gave the following error.
    Type '{ notification: React.ForwardRefExoticComponent<NotificationSnackProps & React.RefAttributes<HTMLDivElement>>; otherNotifications: React.ForwardRefExoticComponent<...>; }' is not assignable to type '{ default?: JSXElementConstructor<any> | undefined; error?: JSXElementConstructor<any> | undefined; success?: JSXElementConstructor<any> | undefined; warning?: JSXElementConstructor<...> | undefined; info?: JSXElementConstructor<...> | undefined; }'.
      Object literal may only specify known properties, and 'notification' does not exist in type '{ default?: JSXElementConstructor<any> | undefined; error?: JSXElementConstructor<any> | undefined; success?: JSXElementConstructor<any> | undefined; warning?: JSXElementConstructor<...> | undefined; info?: JSXElementConstructor<...> | undefined; }'.
  Overload 2 of 2, '(props: SnackbarProviderProps, context: any): SnackbarProvider', gave the following error.
    Type '{ notification: React.ForwardRefExoticComponent<NotificationSnackProps & React.RefAttributes<HTMLDivElement>>; otherNotifications: React.ForwardRefExoticComponent<...>; }' is not assignable to type '{ default?: JSXElementConstructor<any> | undefined; error?: JSXElementConstructor<any> | undefined; success?: JSXElementConstructor<any> | undefined; warning?: JSXElementConstructor<...> | undefined; info?: JSXElementConstructor<...> | undefined; }'.
      Object literal may only specify known properties, and 'notification' does not exist in type '{ default?: JSXElementConstructor<any> | undefined; error?: JSXElementConstructor<any> | undefined; success?: JSXElementConstructor<any> | undefined; warning?: JSXElementConstructor<...> | undefined; info?: JSXElementConstructor<...> | undefined; }'.

252               notification: NotificationSnack,
```

**Comment by codeBelt** [comment]
I had a similar issue because I put `notistack.d.ts` outside of my `types` folder.

This might be a work around if needed:

```ts
import {VariantOverrides as NotistackVariantOverrides} from "notistack";

/**
 * @link https://notistack.com/features/customization
 * @link https://graffino.com/til/how-to-extend-interfaces-declared-in-external-libraries-in-typescript
 */
declare module 'notistack' {
    export interface VariantOverrides extends NotistackVariantOverrides {
        survey: true;
    }
}
```

**Comment by marc-at-brightnight** [comment]
@codeBelt thanks, although that doesn't seem to help. It's merely extending an empty interface so not sure I understand the concept there.

I moved around my `notistack.d.ts` to outside my `src/` folder to the same level as my `tsconfig.json` but no improvement. I tried messing with my `tsconfig.json` to have different variations of `include` as well, didn't seem to work. I would think `"include": ["src/*.d.ts", "src"]` should work.

Not sure if it matters but I'm working in a monorepo, where the changes I am making are inside my `ui/src/` folder, i.e. `ui/src/notistack.ts` and `ui/src/App.tsx`, etc.

```
packages
├── common
├── server
└── ui
```
